[quant][graphmode] Refactor fusion to use the new Pattern format

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69335
* __->__ #68770
* #68769
* #68229

Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

jerryzh168 committed:
[quant][graphmode] Refactor fusion to use the new Pattern format

Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

vkuzo commented on torch/ao/quantization/fuser_method_mappings.py:
```diff
@@ -104,7 +106,7 @@ def fuse_linear_bn(linear, bn):
         return nn.utils.fusion.fuse_linear_bn_eval(linear, bn)
 
-DEFAULT_OP_LIST_TO_FUSER_METHOD : Dict[Tuple, Union[nn.Sequential, Callable]] = {
+DEFAULT_OP_LIST_TO_FUSER_METHOD: Dict[Tuple, Union[nn.Sequential, Callable]] = {
```
this one is easier to use for most common use cases today.  Thoughts about keeping the simple syntax for linear patterns, and enabling the more complicated syntax for subgraphs?  If we deprecate the simple syntax, we force everyone to use the harder syntax even if their use cases might be ok with the current syntax.

jerryzh168 replied:
the difference with current syntax is only the order, do you feel it's worthwhile to do it?

vkuzo replied:
I thought it is the order as well as the writing the sequence as a graph, i.e.

```
# before
Conv, BN, ReLU

# after
(ReLU, (BN, Conv))
```

is there a downside in allowing the user to specify `Conv, BN, ReLU`, and doing the rewrite for them without the user having to care about implementation details?

jerryzh168 replied:
I think the downside is introducing two format for people to define the patterns, and people need to think about whether they need to use the simple format or the complex format

jerryzh168 replied:
also the user is backend developers, not general public users. how about using this for now and we can add a simplified layer later if needed? It feels a bit overkill to optimize for these now since not sure how often these will be used and how much of a burden this would be for users.

vkuzo replied:
> also the user is backend developers, not general public users

the people who work on the code (you, myself, our team members) are also an important audience, as we are the ones who will have to work on this code in the future.  Would you be interested in doing an informal poll of our teammates on which format they find easier and by how much?

> I think the downside is introducing two format for people to define the patterns, and people need to think about whether they need to use the simple format or the complex format

right, seems like the use cases committed to the codebase today are 100% expressible by the simple format.  Is it really worth breaking the mental model for these?  Can we make the simple experience easy to understand, and only require the users with the more complex use cases to understand the complex syntax?

vkuzo replied:
we chatted offline, the consensus was to add a simple syntax option in a follow-up PR

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).

jerryzh168 committed:
Update on "[quant][graphmode] Refactor fusion to use the new Pattern format"


Summary:
Previous fusion only works for a sequnce of ops, which is not general enough for fusion patterns
that is defined by a subgraph, this PR refactors that to make it more general

Test Plan:
```
python test/test_quantization.py TestFuseFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32602637](https://our.internmc.facebook.com/intern/diff/D32602637)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32602637).
